go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+			BucketName:      &bs.bucketName,
+			ObjectName:      &absKey,
+			UploadId:        startRes.UploadId,
+			RequestMetadata: common.RequestMetadata{},
+		})
+		return "", err
+	}
+
+	commitReq := objectstorage.CommitMultipartUploadRequest{
+		NamespaceName:                &namespace,
+		BucketName:                   &bs.bucketName,
+		ObjectName:                   &absKey,
+		UploadId:                     startRes.UploadId,
+		CommitMultipartUploadDetails: objectstorage.CommitMultipartUploadDetails{PartsToCommit: parts},
+	}
+
+	if expectedVersion != "" {
+		commitReq.IfMatch = &expectedVersion
+	} else {
+		commitReq.IfNoneMatch = &star
+	}
+
+	commitRes, err := bs.client.CommitMultipartUpload(ctx, commitReq)
+	if err != nil {
+		return "", err
+	}
+
+	return fmtstr(commitRes.ETag), nil
+}
+
+func (bs *OCIBlobstore) uploadParts(ctx context.Context, objectName, uploadID string, numParts int, totalSize int64, reader io.Reader) ([]objectstorage.CommitMultipartUploadPartDetails, error) {
+	return uploadParts(ctx, objectName, uploadID, numParts, bs.concurrentListeners, totalSize, defaultBatchSize, reader, bs.uploadPart)
+}
+
+func (bs *OCIBlobstore) uploadPart(ctx context.Context, objectName, uploadID string, partNumber int, contentLength int64, reader io.Reader) (objectstorage.CommitMultipartUploadPartDetails, error) {
+	if objectName == "" {
+		return objectstorage.CommitMultipartUploadPartDetails{}, errors.New("object name required to upload part")
+	}
+
+	if uploadID == "" {
+		return objectstorage.CommitMultipartUploadPartDetails{}, errors.New("upload id required to upload part")
+	}
+
+	namespace, err := bs.getNamespace(ctx)
+	if err != nil {
+		return objectstorage.CommitMultipartUploadPartDetails{}, err
+	}
+
+	res, err := bs.client.UploadPart(ctx, objectstorage.UploadPartRequest{
+		NamespaceName:  &namespace,
+		BucketName:     &bs.bucketName,
+		ObjectName:     &objectName,
+		UploadId:       &uploadID,
+		UploadPartNum:  &partNumber,
+		ContentLength:  &contentLength,
+		UploadPartBody: io.NopCloser(reader),
+	})
+	if err != nil {
+		return objectstorage.CommitMultipartUploadPartDetails{}, err
+	}
+
+	return objectstorage.CommitMultipartUploadPartDetails{
+		Etag:    res.ETag,
+		PartNum: &partNumber,
+	}, nil
+}
+
+func (bs *OCIBlobstore) downloadPart(ctx context.Context, objectName, namespace, etag string, partNumber int, br BlobRange) (*toDownload, error) {
+	if objectName == "" {
+		return nil, errors.New("object name required to upload part")
+	}
+
+	req := objectstorage.GetObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &objectName,
+	}
+
+	star := "*"
+	if etag != "" {
+		req.IfMatch = &etag
+	} else {
+		req.IfNoneMatch = &star
+	}
+
+	byteRange := br.asHttpRangeHeader()
+	if byteRange != "" {
+		req.Range = &byteRange
+	}
+
+	res, err := bs.client.GetObject(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Content.Close()
+
+	p, err := createTempFile()
```

Comment:
I don't think this should be structured to use temporary files and a final copy. Instead, I think:

1) the destination path should be opened for with `os.Create()` before we start calling downloadPart.
2) each `downloadPart` invocation should be given an `io.Writer` to dump its body into.
3) That `io.Writer` instance should be `io.NewOffsetWriter()` constructed against the `*os.File` and the total offset of the range which is being downloaded.

Does that make sense? This works on macOS, Linux and Windows AFAICT at least...

Reply:
havent been able to get this to work